server/src/task_initializer.ts

```diff
@@ -0,0 +1,190 @@
+import createDebug from "debug";
+import { List, Set } from 'immutable'
+import { createHash } from 'node:crypto'
+import fs from 'node:fs/promises'
+import tf from '@tensorflow/tfjs'
+import '@tensorflow/tfjs-node'
+
+import {
+  Task, Digest, TaskProvider, isTask,
+  serialization, models, Model
+} from '@epfml/discojs'
+import type { EncodedModel } from '@epfml/discojs'
+
+const debug = createDebug("server:task_initializer");
+
+/**
+ * The TaskInitializer essentially handles initializing a Task and 
+ * its associated EncodedModel.
+ * 
+ * We rely on a TaskInitializer to abstract the (asynchronous) logic of getting the model
+ * when not provided.
+ * Depending on the case, getting the model is done by reading the model files
+ * from disk if they exists, downloading them from a URL or 
+ * initializing the model from its architecture definition. 
+ *  
+ * We work with EncodedModels rather than Models because they are sent encoded 
+ * to clients. Since the server doesn't need to use the Model, we
+ * simply leave it already encoded and ready to be sent to clients
+ * 
+ * Due to its asynchronous nature, TaskInitializer acts like an EventEmitter, 
+ * by registering callbacks on new tasks and emitting a 'newTask' event 
+ * when a new task has been added.
+ * 
+ * Tasks are usually passed to TaskInitializer when booting the server
+ * and objects depending on tasks and models can subscribe to 
+ * the 'newTask' event to run callbacks whenever a new Task and EncodedModel are initialized.
+ */
+export class TaskInitializer {
+  // Keep track of previously initialized task-model pairs
+  #initializedTasks = Set<[Task, EncodedModel]>()
+  // List of callback to apply to future task-model pairs added
+  private listeners = List<(t: Task, m: EncodedModel) => Promise<void>>()
+
+  // Register a callback to be ran on all tasks
+  on(_: 'newTask', callback: (t: Task, m: EncodedModel) => Promise<void>): void {
+    // Apply the callback to already initialized task-model pairs
+    this.#initializedTasks.forEach(async ([t, m]) => { await callback(t, m) })
+    // Register the callback that will be ran when new tasks are added
+    this.listeners = this.listeners.push(callback)
+  }
+
+  // Emit a 'newTask' event, 
+  // It runs all the registered callbacks with the new task and model
+  emit(_: 'newTask', task: Task, model: EncodedModel): void {
```

Comment:
evetns can should only be emitted via `addTask`

```suggestion
  #emit(_: 'newTask', task: Task, model: EncodedModel): void {
```